### PR TITLE
Script to create single Q file from model

### PR DESCRIPTION
**Pull request by jpa198**
Script to create both Q files for a run
Script to create all Q files for a run
Tidy up fault damage zone with new VMF class

### CODE REVIEW

File: srf_generation/velocity_model_generation/q_model_generation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import argparse
+
+
+def load_args():
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/__init__.py:4]: It's kinda weird having this in __init__

If it's necessary, a comment explaining why would be good.

File: srf_generation/velocity_model_generation/q_model_generation/generate_all_qp_qs.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+
+def generate_qp_qs(args: argparse.Namespace, fault: str):
+    vm_params_path = simulation_structure.get_vm_params_path(
+        args.cybershake_root, fault
+    )
+    vm_params = load_yaml(vm_params_path)
+
+    xys = generate_xys(vm_params)
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/generate_all_qp_qs.py:30]: xy_points? a better descriptive variable / function may help?

File: srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        vm_params["nx"],
+        vm_params["ny"],
+        vm_params["hh"],
+    )
+    # Convert the lat lon pairs for each point into grid points in the model domain
+    xys = ll2xy(np.asarray(lls), Q_LLinv)
+    return xys
+
+
+def get_interp(infile: str):
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py:65]: get_interp_grid

File: srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    # Get the interpolator from the model
+    interpolator = get_interp(model)
+
+    # Set the step size from a rudimentary ram use calculation
+    step_size = int(np.floor(MEMORY / (240 * nx * ny)))
+    if step_size < 1:
+        # If we don't have enough ram to do this,
+        # then raise an exception and allow the user to increase ram allocation or reduce domain size
+        raise MemoryError(
+            "Not enough ram to perform this operation. Either allocate more ram, or reduce domain size"
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py:94]: Memory should be argparsable. If your option is to allocate more RAM then you don't want to edit a code file?

File: srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+
+def main():
+    args = load_args()
+
+    vm_params = load_yaml(args.vm_params_path)
+
+    xys = generate_xys(vm_params)
+
+    print("Generating first q file")
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/generate_qp_qs.py:179]: qp? Rather than saying first?

**Review by jasonmotha — CHANGES_REQUESTED** [review]

File: srf_generation/velocity_model_generation/q_model_generation/__init__.py (hunk 1 above)
Comment (jpa198) [reply srf_generation/velocity_model_generation/q_model_generation/__init__.py:4]: Huh, that's an artifact of my default template...
Will remove

File: srf_generation/velocity_model_generation/q_model_generation/generate_all_qp_qs.py [hunk 6]
```diff
@@ -70,6 +70,14 @@ def load_args():
         help="Path to the fault selection file.",
     )
 
+    parser.add_argument(
+        "-n",
+        "--n_processes",
+        type=int,
+        help="Number of Qp and Qs files to generate at once. At most one per fault/event.",
```
Comment (jasonmotha) [inline srf_generation/velocity_model_generation/q_model_generation/generate_all_qp_qs.py:77]: Per rupture? Or per fault? Or per realisation?

Comment (jpa198) [reply srf_generation/velocity_model_generation/q_model_generation/generate_all_qp_qs.py]: As the comment says, per fault.

**Review by jasonmotha — DISMISSED** [review]

**Review by jasonmotha — DISMISSED** [review]

**Review by jasonmotha — APPROVED** [review]

File: srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+def load_args():
+    parser = argparse.ArgumentParser(allow_abbrev=False)
+    parser.add_argument("infile", type=abspath)
+    parser.add_argument("outfile", type=abspath)
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    args = load_args()
```
Comment (claudio525) [inline srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py:18]: I think this part belongs under `__main__`, not in the main function.To me main is a function like any other (`__main__` is the starting point in  a script), hence it should be re-usable and self-contained. So I usually do all the argument stuff under` __main__` and then pass the parsed arguments to main, i.e. main becomes
```
def main(infile: Path, outfile: Path):
    .....
```

Anyways just my 5 cents, in my opinion it makes the code cleaner, so I like it haha.

File: srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    # load model data
+    data = pd.read_csv(args.infile, sep="\s+", skiprows=1, header=0, index_col=None)
+    data.columns = ["vals", "SF", "x", "y", "z", "lat", "lon"]
+    data.sort_values(["x", "y", "z"], axis=0, inplace=True)
+    data.reset_index(inplace=True, drop=True)
+    xs = data["x"].unique()
+    ys = data["y"].unique()
+    zs = data["z"].unique()
+    xs_count = len(xs)
```
Comment (claudio525) [inline srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py:28]: What about?
```
xs_count, ys_count, zs_count = len(xs), len(ys), len(zs)
```
Especially for such simple repeated variable definitions?

**Review by claudio525 — APPROVED** [review]

File: srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py (hunk 7 above)
Comment (jasonmotha) [reply srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py:18]: My 2c is that I don't like args (and really anything unnecessary) as a global variable so returning args into __main__ gives it global scope.

I normally do mains like this, but only use it for processing arguments really and then hands off to a function that does the body of the script. So I'd only use main in the script that has it.

I definitely see why you would do it the way you do. I think we might have even had this conversation before at some point.

File: srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py (hunk 8 above)
Comment (jpa198) [reply srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py:28]: Huh, I don't usually think of putting that sort of thing on one line
I normally only use that notation for tuple/list packing/unpacking
I could change it, but you'd have to approve it again

File: srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py (hunk 7 above)
Comment (jpa198) [reply srf_generation/velocity_model_generation/q_model_generation/EP_to_bin.py:18]: I think Jason got to me first on `main()` vs `__main__` style, so I normally go with his way.
Using non constant globals and args outside main is out, `__name__=="__main__":main()` is in :stuck_out_tongue:

**Pull request merged**